pkg/service/replay/utils.go

```diff
@@ -65,3 +78,487 @@ func (t *testUtils) SimulateRequest(ctx context.Context, _ uint64, tc *models.Te
 	}
 	return nil, nil
 }
+
+func CalculateAndInsertTestCoverage(ctx context.Context, logger *zap.Logger, reportDB ReportDB, testRunID string, language config.Language) {
+	var coverageData models.TestCoverage
+	logger.Info("calculating coverage for the test run and inserting it into the report")
+	var err error
+	switch language {
+	case models.Go:
+		coverageData, err = CalGoCoverage()
+	case models.Python:
+		coverageData, err = CalPythonCoverage(ctx)
+	case models.Node:
+		coverageData, err = CalTypescriptCoverage(ctx)
+	case models.Java:
+		coverageData, err = CalJavaCoverage(ctx)
+	}
+	if err != nil {
+		utils.LogError(logger, err, "failed to calculate coverage for the test run")
+	} else {
+		err = reportDB.InsertCoverageReport(ctx, testRunID, &coverageData)
+		if err != nil {
+			utils.LogError(logger, err, "failed to update report with the coverage data")
+		}
+	}
+}
+
+func downloadAndExtractJaCoCoBinaries(version, dir string) error {
+	cliPath := filepath.Join(dir, "jacococli.jar")
+
+	downloadURL := fmt.Sprintf("https://github.com/jacoco/jacoco/releases/download/v%s/jacoco-%s.zip", version, version)
+
+	_, err := os.Stat(cliPath)
+	if err == nil {
+		return nil
+	}
+
+	resp, err := http.Get(downloadURL)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	zipReader, err := zip.NewReader(bytes.NewReader(body), int64(len(body)))
+	if err != nil {
+		return err
+	}
+
+	for _, file := range zipReader.File {
+		if strings.HasSuffix(file.Name, "jacococli.jar") {
+			cliFile, err := file.Open()
+			if err != nil {
+				return err
+			}
+			defer cliFile.Close()
+
+			outFile, err := os.Create(cliPath)
+			if err != nil {
+				return err
+			}
+			defer outFile.Close()
+
+			_, err = io.Copy(outFile, cliFile)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	cliStat, err := os.Stat(cliPath)
+
+	if os.IsNotExist(err) || cliStat != nil {
+		return fmt.Errorf("failed to find JaCoCo binaries in the distribution")
+	}
+
+	return nil
+}
+
+func mergeJacocoCoverageFiles(ctx context.Context, jacocoCliPath string) error {
+	// Find all .exec files starting with "test-set" in the target directory
+	sourceFiles, err := filepath.Glob("target/test-set*.exec")
+	if err != nil {
+		return fmt.Errorf("error finding coverage files: %w", err)
+	}
+	if len(sourceFiles) == 0 {
+		return errors.New("no coverage files found")
+	}
+
+	// Construct the command arguments
+	args := []string{
+		"java",
+		"-jar",
+		jacocoCliPath,
+		"merge",
+	}
+
+	// Append each source file to the command
+	for _, file := range sourceFiles {
+		args = append(args, file)
+	}
+
+	// Specify the output file
+	args = append(args, "--destfile", "target/keploy-e2e.exec")
+
+	cmd := exec.CommandContext(ctx, args[0], args[1:]...)
```

Comment:
make sure we dont show these logs to user, because all we want for them to see is application and keploy logs may be print them as debug logs

Reply:
should I change all current error logs introduced in this PR which are related to coverage thing to Warn ?

Reply:
yes and the success logs should not be shown like if you are mergin the jacoco exec files I dont want user to see the logs because they are irrelevant to the user.

Reply:
Should I print a warn statement saying to check debug logs if skipcoverage is set to true by keploy?

Reply:
No if there is an error in any coverage related thing log that as warn and if there is any logs from the commands you are running like jaoco merge or report dont show the logs to user log them as debug thats it

Reply:
done